zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/OperationReferenceTest.java

```diff
@@ -0,0 +1,94 @@
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.camunda.zeebe.qa.util.cluster.TestStandaloneBroker;
+import io.camunda.zeebe.qa.util.junit.ZeebeIntegration;
+import io.camunda.zeebe.qa.util.junit.ZeebeIntegration.TestZeebe;
+import io.camunda.zeebe.test.util.junit.AutoCloseResources.AutoCloseResource;
+import io.camunda.zeebe.test.util.record.ProcessInstanceRecordStream;
+import io.camunda.zeebe.test.util.record.RecordingExporter;
+import java.time.Duration;
+import org.junit.jupiter.api.Test;
+
+@ZeebeIntegration
+public class OperationReferenceTest {
+  private static final long OPERATION_REFERENCE = 1234L;
+
+  @TestZeebe
+  private static final TestStandaloneBroker ZEEBE =
+      new TestStandaloneBroker().withRecordingExporter(true);
+
+  @AutoCloseResource
+  private final ZeebeClient client =
+      ZEEBE.newClientBuilder().defaultRequestTimeout(Duration.ofMinutes(2)).build();
+
+  @Test
+  void shouldIncludeOperationReferenceInExportedCommandRecord() {
+    // Given
+    final ZeebeResourcesHelper helper = new ZeebeResourcesHelper(client);
+    final var modelInstance = helper.createSingleJobModelInstance("test", c -> {});
+    final long processDefinitionKey = helper.deployProcess(modelInstance);
+    final long processInstanceKey = helper.createProcessInstance(processDefinitionKey);
+
+    // When
+    client
+        .newCancelInstanceCommand(processInstanceKey)
+        .operationReference(OPERATION_REFERENCE)
+        .send()
+        .join();
+
+    // Then
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.CANCEL)
+                .withRecordKey(processInstanceKey)
+                .limit(1)
+                .getFirst()
+                .getOperationReference())
+        .describedAs("Should contain client operationReference")
+        .isEqualTo(OPERATION_REFERENCE);
+  }
+
+  @Test
+  void shouldIncludeOperationReferenceInFollowUpRecords() {
+    // Given
+    final ZeebeResourcesHelper helper = new ZeebeResourcesHelper(client);
+    final var modelInstance = helper.createSingleJobModelInstance("test", c -> {});
+    final long processDefinitionKey = helper.deployProcess(modelInstance);
+    final long processInstanceKey = helper.createProcessInstance(processDefinitionKey);
+
+    // When
+    client
+        .newCancelInstanceCommand(processInstanceKey)
+        .operationReference(OPERATION_REFERENCE)
+        .send()
+        .join();
+
+    // Then
+    final var cancelCommand =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.CANCEL)
+            .withRecordKey(processInstanceKey)
+            .limit(1)
+            .getFirst();
+
+    final ProcessInstanceRecordStream followUpRecords =
+        RecordingExporter.processInstanceRecords()
+            .withProcessInstanceKey(processInstanceKey)
+            .withSourceRecordPosition(cancelCommand.getPosition());
+
+    assertThat(followUpRecords)
```

Comment:
❌ This only verifies the records of `ValueType` `PROCESS_INSTANCE`, but all follow-up records must have the operation reference. Please use `RecordingExporter.records()` instead and don't filter it by `processInstanceKey`.

❌ This stream is not short-circuited and will take 5 seconds to evaluate. Please limit the stream or short-circuit it some other way.

Reply:
>❌ This only verifies the records of ValueType PROCESS_INSTANCE, but all follow-up records must have the operation reference. Please use RecordingExporter.records() instead and don't filter it by processInstanceKey.

good point, will update the tests! Thank you :orange_heart:

>❌ This stream is not short-circuited and will take 5 seconds to evaluate. Please limit the stream or short-circuit it some other way.

I tried `asList`, dynamic limiting until source position > cancel command position, etc. Nothing helped the time. I also thought `all match` in the assert should be short-circuiting

Reply:
Can't we use a similar trick? i.e. stop when you see the terminate record? Just a suggestion :shrug: